### PR TITLE
Add Hugepage-Isolator

### PR DESCRIPTION
**Pull request by swatisehgal**
+ support for Hugetlbfs Cgroup in Kubelet

Co-Authored-By: Louise Daly louise.m.daly@intel.com

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Review by ConnorDoyle — CHANGES_REQUESTED** [review]
Could you please remove the following from the patch?

* Duplicated isolator library files (import the existing one instead?)
* The hugepage-isolator executable (also add to .gitignore in that directory)

**Comment by lmdaly** [comment]
Hey @ConnorDoyle made those changes. Thanks for the review!

File: cluster/addons/hp-client/examples/example-hugepage-pod.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    pod.alpha.kubernetes.io/isolation-api: '{
+        "hugepage": "2MB"
+    }'
+  labels:
+    iso-resource: hugepage
+  name: nginxpod
+spec:
+  containers:
+  - name: nginxcontainer
+    image: si-docker.ir.intel.com/ssehgal/mmap
```
Comment (squall0gd) [inline cluster/addons/hp-client/examples/example-hugepage-pod.yaml]: Is this image is visible/available outside of Intel? As far as we want to upstream it, community won't be able to use your example ;)

File: cluster/addons/hp-client/examples/example-hugepage-pod.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    command: [ "/bin/bash", "-c", "--" ]
+    args: [ "while true; do sleep 300000; done;" ]
```
Comment (squall0gd) [inline cluster/addons/hp-client/examples/example-hugepage-pod.yaml]: You can prepare your image like this:
```docker
...
ENTRYPOINT ["/bin/bash", "-c", "--", "sleep"]
CMD "inf"
```

Your will be able run this image w/o any command and args. If you need to change amount of time, then you can do it with command, by passing time like this:

```yaml
    command: 300
```

File: cluster/addons/hp-client/examples/hugepage-client.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      labels:
+        isolator: hugepage
+    spec:
+      containers:
+      - command:
+        - /hugepage-isolator
+        - -v
+        - "10"
+        - -logtostderr
+        image: si-docker.ir.intel.com/lmdaly/cgroup-hugepage-isolator
```
Comment (squall0gd) [inline cluster/addons/hp-client/examples/hugepage-client.yaml]: It's better to use `cgroup-hugepage-isolator` here(which is declared in Dockerfile)

File: cluster/addons/hp-client/hugepages/discovery/hp_discovery.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+                return -1, err
+        }
+        hpTotal := bytes.TrimSpace(totalBytes)
+        hpTotalNum, err := strconv.Atoi(string(hpTotal))
+        if err != nil {
+                return -1, err
+        }
+        return hpTotalNum, nil
+}
+
```
Comment (squall0gd) [inline cluster/addons/hp-client/hugepages/discovery/hp_discovery.go]: remove this line

File: cluster/addons/hp-client/hugepage-isolator.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+const (
+	// kubelet eventDispatcher address
+	eventDispatcherAddress = "localhost:5433"
+	// iso-client own address
+	isolatorLocalAddress = "localhost:5444"
+	// name of isolator
+	name = "cgroup-hugetlb-limit"
+)
+
+// TODO: split it to smaller functions
```
Comment (squall0gd) [inline cluster/addons/hp-client/hugepage-isolator.go]: is it still valid?

File: cluster/addons/hp-client/hugepage-isolator.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/cluster/addons/hp-client/hugepages"
+	"k8s.io/kubernetes/cluster/addons/iso-client/isolator"
+)
+
+const (
+	// kubelet eventDispatcher address
+	eventDispatcherAddress = "localhost:5433"
+	// iso-client own address
+	isolatorLocalAddress = "localhost:5444"
```
Comment (squall0gd) [inline cluster/addons/hp-client/hugepage-isolator.go]: Can you change your `isolatorLocalAddress` to sth different? It is overlapping with already existing `coreaffinity` isolator.

File: cluster/addons/hp-client/hugepages/discovery/hp_discovery.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package discovery
+
+import (
+	"bytes"
+	"io/ioutil"
+	"strconv"
+)
+
+func DiscoverHugepageNumber() (int, error) {
+        procFsPath := "/proc/sys/vm/nr_hugepages"
```
Comment (squall0gd) [inline cluster/addons/hp-client/hugepages/discovery/hp_discovery.go]: s/procFsPath/hugepageInfoPath/

File: cluster/addons/hp-client/hugepages/hpIsolator.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	if pod.Annotations["pod.alpha.kubernetes.io/isolation-api"] != "" {
+		spec := &isoSpec{}
+		err := json.Unmarshal([]byte(pod.Annotations["pod.alpha.kubernetes.io/isolation-api"]), spec)
+		if err != nil {
+			glog.Fatalf("Cannot unmarshal isoSpec: %v", err)
+			return "", err
+		}
+		if spec.Hugepage != "" {
+			return spec.Hugepage, nil
+		}
```
Comment (squall0gd) [inline cluster/addons/hp-client/hugepages/hpIsolator.go]: you can move annotations parsing to another function, so each of your function will be responsible for one task :)

File: cluster/addons/hp-client/hugepages/hpIsolator.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	
+	cgroupResource := []*lifecycle.IsolationControl{
+		{
+			MapValue: hugePageMapValue,
+			Value: cgroupResourceValue,
+			Kind:  lifecycle.IsolationControl_CGROUP_HUGETLB_LIMIT,
+		},
+	}
+	
+	return cgroupResource, nil
```
Comment (squall0gd) [inline cluster/addons/hp-client/hugepages/hpIsolator.go:151]: probably you should split logic here to other functions.

**Pull request closed without merging**